src/server/qgsconfigcache.cpp

```diff
@@ -152,25 +187,128 @@ QDomDocument *QgsConfigCache::xmlDocument( const QString &filePath )
       return nullptr;
     }
     mXmlDocumentCache.insert( filePath, xmlDoc );
-    mFileSystemWatcher.addPath( filePath );
     xmlDoc = mXmlDocumentCache.object( filePath );
     Q_ASSERT( xmlDoc );
   }
   return xmlDoc;
 }
 
-void QgsConfigCache::removeChangedEntry( const QString &path )
+
+void QgsConfigCache::cacheProject( const QString &path, QgsProject *project )
+{
+  mProjectCache.insert( path, new std::pair<QDateTime, std::unique_ptr<QgsProject> >( project->lastModified(),  std::unique_ptr<QgsProject>( project ) ) );
+
+  mStrategy->entryInserted( path );
+}
+
+void QgsConfigCache::removeEntry( const QString &path )
 {
   mProjectCache.remove( path );
 
   //xml document must be removed last, as other config cache destructors may require it
   mXmlDocumentCache.remove( path );
 
+  mStrategy->entryRemoved( path );
+}
+
+// slots
+
+void QgsConfigCache::removeChangedEntry( const QString &path )
+{
+  removeEntry( path );
+}
+
+
+void QgsConfigCache::removeChangedEntries()
+{
+  // QCache::keys returns a QList so it is safe
+  // to mutate while iterating
+  for ( const auto &path : mProjectCache.keys() )
```

Comment:
take a const copy before iterating

Reply:
I'm confused, a copy wont mutate the original, why make it const ?

Reply:
Afaik, the keys() method returns already a copy so it should be safe to use it directly in the iteration.

Reply:
Done, but I'm curious to know why the previous statement was not suitable.

Reply:
See "avoid detach" in https://www.kdab.com/goodbye-q_foreach/ for the technical background